File: SDK Examples/Examples/C#_Console/Counter_ContinueCompare/ContinueCompare.cs
/*******************************************************************************
Copyright (c) 1983-2016 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*     ContinueCompare.cs
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use  Continue Compare Counter function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*    3  Set the 'channelStart' as the start channel of the counter to operate
*	  4  Set the 'channelCount' as the channel count of the counter to operate.
*    5  Set the 'comValueTab' as the comparison value table.
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Reflection;
using System.Runtime.InteropServices;
using Automation.BDaq;

namespace Counter_UpDownCounter
{
   class UpDownCounter
   {
      static int[][] comValueTab = new int[2][];

      static void Main(string[] args)
      {
         //-----------------------------------------------------------------------------------
         // Configure the following parameters before running the demo
         //-----------------------------------------------------------------------------------
			string deviceDescription = "PCI-1784,BID#0";
         string profilePath = "../../../profile/PCI-1784.xml";
			int channelStart = 0;
			int channelCount = 1;

         // initialize the compare value table.
         comValueTab[0] = new int []{ 500, 3125, 8520 };
         comValueTab[1] = new int []{ 10000, 13040, 17550 };

			ErrorCode errorCode = ErrorCode.Success;

         // Step 1: Create a 'UdCounterCtrl' for UpDown Counter function.
         UdCounterCtrl udCounterCtrl = new UdCounterCtrl();

         // Step 2: Set the notification event Handler by which we can known the state of operation effectively.
         udCounterCtrl.UdCntrEvent += new EventHandler<UdCntrEventArgs>(udCounterCtrl_ConCmpValue);

         try
         {
            // Step 3: Select a device by device number or device description and specify the access mode.
            // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
            udCounterCtrl.SelectedDevice = new DeviceInformation(deviceDescription);
				errorCode = udCounterCtrl.LoadProfile(profilePath);//Loads a profile to initialize the device.
            if (BioFailed(errorCode))
            {
               throw new Exception();
            }

            // Step 4: Set necessary parameters.
				udCounterCtrl.ChannelStart = channelStart;
				udCounterCtrl.ChannelCount = channelCount;

            // Step 5: Set counting type for UpDown Counter 
				int channelCountMax = udCounterCtrl.Features.ChannelCountMax;
				for (int i = channelStart; i < channelStart + channelCount; i++)
				{
					udCounterCtrl.Channels[i % channelCountMax].CountingType = CountingType.AbPhaseX1;
				}

            //Step 6: Set compare table
            udCounterCtrl.CompareSetTable(channelStart, comValueTab[0].Length, (int[])comValueTab[0].Clone());

            // Step 7: Start UpDown Counter 
            udCounterCtrl.Enabled = true;

            // Step 8: Read counting value: connect the input signal to channels you selected to get event counter value.
            Console.WriteLine(" UpDown Counter is in progress...\nconnect the input signal to");
            Console.WriteLine(" any key to quit!\n\n");
            while (!Console.KeyAvailable)
            {
               Thread.Sleep(1000);
               int [] buffer = {0};
               udCounterCtrl.Read(1, buffer);
               Console.WriteLine("channel {0} Current  counts  :{1}", channelStart, buffer[0]);
            }

            // Step 9: stop UpDown Counter
            udCounterCtrl.CompareClear(channelStart);
            udCounterCtrl.Enabled = false;
         }
         catch(Exception e)
         {
            // Something is wrong
            string errStr = BioFailed(errorCode) ? " Some error occurred. And the last error code is " + errorCode.ToString()
                                          : e.Message;
            Console.WriteLine(errStr);
         }
         finally
         {
            // Step 10: Close device and release any allocated resource.
            udCounterCtrl.Dispose();
            Console.ReadKey(false);
         }
      }

      static int conCmpOccursCount = 0;
      static int tabIndex = 0;
      static EventId[] evntID = new EventId[]{EventId.EvtCntCompareTableEnd0,EventId.EvtCntCompareTableEnd1,EventId.EvtCntCompareTableEnd2,EventId.EvtCntCompareTableEnd3,
                                  EventId.EvtCntCompareTableEnd4,EventId.EvtCntCompareTableEnd5,EventId.EvtCntCompareTableEnd6,EventId.EvtCntCompareTableEnd7};
      static EventId[] evntCompID = new EventId[]{EventId.EvtCntPatternMatch0,EventId.EvtCntPatternMatch1,EventId.EvtCntPatternMatch2,EventId.EvtCntPatternMatch3,
                                          EventId.EvtCntPatternMatch4,EventId.EvtCntPatternMatch5,EventId.EvtCntPatternMatch6,EventId.EvtCntPatternMatch7};

      static void udCounterCtrl_ConCmpValue(object sender, UdCntrEventArgs e)
      {
         int channel = ((UdCounterCtrl)sender).ChannelStart;
         UdCounterCtrl udCounterCtrl = (UdCounterCtrl)sender;
         if (evntCompID[channel] == e.Id || evntID[channel] == e.Id)
         {
            Console.WriteLine("Channel {0} Compare occurs {1} time(times)", channel, ++conCmpOccursCount);
            Console.WriteLine("Compare value is {0}.\n", comValueTab[tabIndex % 2][(conCmpOccursCount-1) % 3]);
         }
         // Change the compare value table.
         if (evntID[channel] == e.Id)
         {
            tabIndex++;
            udCounterCtrl.CompareSetTable(channel, comValueTab[tabIndex % 2].Length, (int[])comValueTab[tabIndex % 2].Clone());
         } 
      }

      static bool BioFailed(ErrorCode err)
      {
         return err < ErrorCode.Success && err >= ErrorCode.ErrorHandleNotValid;
      }
   }
}

File: SDK Examples/Examples/C++_Console/Counter_ContinueCompare/ContinueCompare.cpp
/*******************************************************************************
Copyright (c) 1983-2016 Advantech Co., Ltd.
********************************************************************************
THIS IS AN UNPUBLISHED WORK CONTAINING CONFIDENTIAL AND PROPRIETARY INFORMATION
WHICH IS THE PROPERTY OF ADVANTECH CORP., ANY DISCLOSURE, USE, OR REPRODUCTION,
WITHOUT WRITTEN AUTHORIZATION FROM ADVANTECH CORP., IS STRICTLY PROHIBITED. 
================================================================================
REVISION HISTORY
--------------------------------------------------------------------------------
$Log: $
--------------------------------------------------------------------------------
$NoKeywords:  $
*/
/******************************************************************************
*
* Windows Example:
*     ContinueCompare.cpp
*
* Example Category:
*    Counter
*
* Description:
*    This example demonstrates how to use Continue Compare Counter function.
*
* Instructions for Running:
*    1  Set the 'deviceDescription' for opening the device. 
*	  2  Set the 'profilePath' to save the profile path of being initialized device. 
*    3  Set the 'channelStart' as the start channel of the counter to operate
*	  4  Set the 'channelCount' as the channel count of the counter to operate.
*
* I/O Connections Overview:
*    Please refer to your hardware reference manual.
*
******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "../inc/compatibility.h"
#include "../../../inc/bdaqctrl.h"
using namespace Automation::BDaq;
//-----------------------------------------------------------------------------------
// Configure the following parameters before running the demo
//-----------------------------------------------------------------------------------
#define     deviceDescription L"PCI-1784,BID#0"
const wchar_t* profilePath = L"../../profile/PCI-1784.xml";
int32       channelStart = 0;
int32       channelCount = 1;

inline void waitAnyKey()
{
   do{SLEEP(1);} while(!kbhit());
} 

int comValueTab[2][3] = {{50, 100, 150},{1000, 1304, 1755}};

int conCmpOccursCount = 0;
int tabIndex = 0;
int const evntID[8] = {EvtCntCompareTableEnd0,EvtCntCompareTableEnd1,EvtCntCompareTableEnd2,EvtCntCompareTableEnd3,
                       EvtCntCompareTableEnd4,EvtCntCompareTableEnd5,EvtCntCompareTableEnd6,EvtCntCompareTableEnd7};
int const evntCompID[8] = {EvtCntPatternMatch0,EvtCntPatternMatch1,EvtCntPatternMatch2,EvtCntPatternMatch3,
                           EvtCntPatternMatch4,EvtCntPatternMatch5,EvtCntPatternMatch6,EvtCntPatternMatch7};
// This function is used to deal with 'Cntr' Event.
void BDAQCALL OnCounterEvent(void * sender, UdCntrEventArgs * args, void * userParam)
{
   UdCounterCtrl *udCounterCtrl = (UdCounterCtrl*)sender;
   int channel = udCounterCtrl->getChannelStart();
   if (evntCompID[channel] == args->Id || evntID[channel] == args->Id)
   {
      printf("Channel %d Compare occurs %d time(times)\n", channel, ++conCmpOccursCount);
      printf("Compare value is %d\n\n",comValueTab[tabIndex % 2][(conCmpOccursCount-1) % 3]);
   }
   if (evntID[channel] == args->Id)
   {
      printf("Channel %d Compare end\n", channel);
      tabIndex++;
      udCounterCtrl->CompareSetTable(channel, 3, comValueTab[tabIndex % 2]);
   }
}

int main(int argc, char* argv[])
{
   ErrorCode ret = Success;
   // Step 1: Create a 'UdCounterCtrl' for UpDown Counter function.
   UdCounterCtrl * udCounterCtrl = UdCounterCtrl::Create();

   // Step 2: Set the notification event Handler by which we can known the state of operation effectively.
   udCounterCtrl->addUdCntrEventHandler(OnCounterEvent, NULL);

   do
   {
      // Step 3: Select a device by device number or device description and specify the access mode.
      // in this example we use ModeWrite mode so that we can fully control the device, including configuring, sampling, etc.
      DeviceInformation devInfo(deviceDescription);
      ret = udCounterCtrl->setSelectedDevice(devInfo);
      CHK_RESULT(ret);
      ret = udCounterCtrl->LoadProfile(profilePath);//Loads a profile to initialize the device.
      CHK_RESULT(ret);
      
      // Step 4: Set necessary parameters.
      ret = udCounterCtrl->setChannelStart(channelStart);
      CHK_RESULT(ret);
      ret = udCounterCtrl->setChannelCount(channelCount);
      CHK_RESULT(ret);

      // Step 5: Set counting type for UpDown Counter 
      /******************************************************************************************************************/
		/*In this example, we use the PCIE-1784 and set 'PulseDirection' as the default CountingType.The details see manual.
		/******************************************************************************************************************/
      Array<UdChannel>* udChannel = udCounterCtrl->getChannels();
      for (int32 i = channelStart; i < channelStart + channelCount; i++)
      {
         ret = udChannel->getItem(i).setCountingType(PulseDirection);
			CHK_RESULT(ret);
      }

      //Step 6: Set compare table
      ret = udCounterCtrl->CompareSetTable(udCounterCtrl->getChannelStart(), 3, &comValueTab[0][0]);
      CHK_RESULT(ret);

      // Step 7: Start UpDown Counter 
      ret= udCounterCtrl->setEnabled(true);
      CHK_RESULT(ret);

      // Step 8: Read counting value: connect the input signal to channels you selected to get UpDown counter value.
      printf("UnDown counter is in progress...\nconnect the input signal to\n ");
      printf("any key will stop UnDown counter!\n\n");
      while ( !kbhit())
      {
         SLEEP(1);//get event UpDown count value per second
         int32 value = 0;
         ret = udCounterCtrl->Read(value);
         CHK_RESULT(ret);
         printf("channel %d Current UpDown count: %u \n\n", channelStart, value);
      }
      
      // Step 9: stop UpDown Counter
      ret = udCounterCtrl->CompareClear();
      CHK_RESULT(ret);
      ret = udCounterCtrl->setEnabled(false);
      CHK_RESULT(ret);
   }while(false);

	// Step 10: Close device and release any allocated resource.
	udCounterCtrl->Dispose();

	// If something wrong in this execution, print the error code on screen for tracking.
   if(BioFailed(ret))
   {
      printf("\nSome error occurred. And the last error code is 0x%X.\n", ret);
      waitAnyKey();
   }  
   return 0;
}
